### PR TITLE
nacos source support fetching instances from all namespaces

### PR DESCRIPTION
**Pull request by believening**
Before we fetch the nacos service only in a specific namespace/group, which can be set in this way:
``` yaml
NacosSource:
  Namespace: public
  Group: DEFAULT_GROUP
```
Now we support fetching the nacos service in all group in all namespaces, which can be set in this way:

``` yaml
NacosSource:
  AllNamespaces: true
```

When "AllNamespaces" is enabled, we merge instances with the same service name from different namespaces and groups into a single ServiceEntry.

We use the API `/nacos/v1/ns/catalog/services` for fetching all services in all groups in a specific namespace. This api, which we found in the nacos 1.4.x console, is not documented in the [nacos openAPI](https://nacos.io/zh-cn/docs/open-api.html) documentation, so we cannot guarantee that this feature is compatible with the latest version of nacos in the community.

### CODE REVIEW

File: staging/src/slime.io/slime/modules/meshregistry/pkg/bootstrap/args.go [hunk 1]
```diff
@@ -166,6 +166,8 @@ type NacosSourceArgs struct {
 	// username and password for nacos auth
 	Username string
 	Password string
+	// fetch services from all namespaces
+	AllNamespaces bool
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/meshregistry/pkg/bootstrap/args.go:170]: add comment for not supporting watching mode

File: staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go [hunk 2]
```diff
@@ -83,31 +117,64 @@ func NewClient(urls []string, username, password string, headers map[string]stri
...
+			return url
+		}
+		return url + "?" + query
+	}
+	var lastErr error
+	l := len(c.urls)
+	shift := rand.Intn(l)
+	for i := 0; i < l; i++ {
+		var curBody io.Reader
+		var copiedBody bytes.Buffer
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go:151]: a little too complicated. 

as at least we should cache the whole body for once, we can read it into a buf and then start the call-loop.

File: staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go [hunk 3]
```diff
@@ -121,37 +188,198 @@ func (c *client) call(method string, url string, body io.Reader) ([]byte, error)
...
+}
+
+func (c *client) listServices(namespaceId, groupName string) ([]string, error) {
+	probeServiceResp, err := c.pagingListServices(namespaceId, groupName, 1)
+	if err != nil {
+		return nil, err
+	}
+	doms := probeServiceResp.Doms
+	if probeServiceResp.Count > serviceListPageSize {
+		pageCount := probeServiceResp.Count/serviceListPageSize + 1
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go:240]: I have some doubts that if `Count == 2000` and we get `pageCount == 3` and then whether we'll get a err when `page == 3`?

File: staging/src/slime.io/slime/modules/meshregistry/pkg/bootstrap/args.go (hunk 1 above)
Comment (believening) [reply staging/src/slime.io/slime/modules/meshregistry/pkg/bootstrap/args.go:173]: do it in another task, maybe cleancode.

File: staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go (hunk 2 above)
Comment (believening) [reply staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go]: copied and create new Reader for every call

File: staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go [hunk 4]
```diff
@@ -83,31 +114,65 @@ func NewClient(urls []string, username, password string, headers map[string]stri
...
+func (c *client) call(api string, method string, header map[string]string, queryParam map[string]string, body io.Reader) ([]byte, error) {
+	query := encodeQuery(queryParam)
+	appendUrl := func(url string) string {
+		if query == "" {
+			return url
+		}
+		return url + "?" + query
+	}
+	var lastErr error
+	var bodyContent string
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go:144]: use bytes or bytes.Buf(reset it each time)

Comment (believening) [reply staging/src/slime.io/slime/modules/meshregistry/pkg/source/nacos/httpclient.go]: replaced

**Review by YonkaFang — APPROVED** [review]

**Pull request merged**